pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -187,40 +204,121 @@ func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(namespacedName string)
 	return nil
 }
 
+// Find IP Pools annotated to StatefulSet via direct annotation or Namespace annotation
+func (c *AntreaIPAMController) getIPPoolsForStatefulSet(ss *appsv1.StatefulSet) []string {
+
+	// Inspect pool annotation for the Pods
+	// In order to avoid extra API call in IPAM driver, IPAM annotations are defined
+	// on Pods rather than on StatefulSet
+	annotations, exists := ss.Spec.Template.Annotations[annotation.AntreaIPAMAnnotationKey]
+	if exists {
+		// Stateful Set Pod is annotated with dedicated IP pool
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	// Inspect Namespace
+	namespace, err := c.namespaceLister.Get(ss.Namespace)
+	if err != nil {
+		// Should never happen
+		klog.Errorf("Namespace %s not found for StatefulSet %s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	annotations, exists = namespace.Annotations[annotation.AntreaIPAMAnnotationKey]
+	if exists {
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	return nil
+
+}
+
+// Look for an IP Pool associated with this StatefulSet, either a dedicated one or
+// annotated to the Namespace. If such IP Pool is found, preallocate IPs for the StatefulSet.
+// This function returns error if pool is not found, or allocation fails.
+func (c *AntreaIPAMController) preallocateIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing create notification", "Namespace", ss.Namespace, "StatefulSet", ss.Name)
+
+	ipPools := c.getIPPoolsForStatefulSet(ss)
+
+	if ipPools == nil {
+		// nothing to preallocate
+		return nil
+	}
+
+	if len(ipPools) > 1 {
+		return fmt.Errorf("annotation of multiple IP Pools is not supported")
+	}
+
+	// Only one pool is supported for now. Dual stack support coming in future.
+	ipPoolName := ipPools[0]
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPoolName, c.crdClient, c.ipPoolLister)
+	if err != nil {
+		return fmt.Errorf("failed to find IP Pool %s: %s", ipPoolName, err)
+	}
+
+	size := int(*ss.Spec.Replicas)
+	err = allocator.AllocateStatefulSet(ss.Namespace, ss.Name, size)
```

Comment:
Have we checked no IP allocated for this StatefulSet already?

Reply:
And add comments somewhere to indicate we do not support IPPool changes.

Reply:
AllocateStatefulSet has this check inside: IPs will not be preallocated if this StatefulSet is already present in the pool.

Reply:
And could we also add a comment about this then? It is just easier to understand.

Reply:
Sure, will do

Reply:
We have a note in docs that explains that IPPool annotation can not be changed without recreating the resource: `Note that the IP pool annotation cannot be updated or deleted without recreating
the resource`. We should add this to validation.

Reply:
That is great. But still feel better to add a comment in the code, so reviewers/readers like me can easily understand why we do not check if pool changes or not.

Reply:
Added comment in code